### DISCUSSION TITLE
Libiconv issues.

### DISCUSSION THREAD

**Discussion by Blady-Com** [jralls/gtk-osx-build#109] [Q&A]
The last time I built GTK-OSX, a year ago, it relied on libiconv from macOS.
Now, it builds libiconv 1.17.
Unfortunately, the two API sets are not the compatible : iconv prefix and libiconv prefix.
In GTK-OSX, the libiconv 1.17 with libiconv prefix is surely required.
But, my app needs libiconv with iconv prefix:
```
dyld[43813]: Symbol not found: _iconv
  Referenced from: <08B3066E-0BE2-35B9-B380-45D49FDA0323> /Users/me/2024a/xnadalib-2024/lib/gnatcoll_iconv.relocatable/libgnatcoll_iconv.dylib.25.0w
  Expected in:     <036D06BA-D319-305E-A5E4-A2FF04BF0E52> /Users/me/2024a/xnadalib-2024/lib/libiconv.2.dylib
```
How to make it correct?
Thanks, Pascal.

**Comment by jralls** [comment]
I think the answer is in the [AdaCore docs](https://docs.adacore.com/gnatcoll-docs/iconv.html): Pass --with-iconv=$PREFIX to GNATCOLL's configure so that the header and library match.

**Comment by Blady-Com** [comment]
Thank you for reminding me of old times :-)
I would have liked this to be the case but this documentation has not corresponded for several years.
Configure is no longer provided but replaced by a Python script without any documentation about choosing the correct libiconv.
I opened an [issue](https://github.com/jralls/gtk-osx-build/discussions/109).

**Reply by jralls** [reply]
@Blady-Com Your issue link just points back here. If you mean to turn this into a gtk-osx issue, those live at https://gitlab.gnome.org/GNOME/gtk-osx/-/issues.

**Reply by Blady-Com** [reply]
Argggh, wrong paste, the right link: [https://github.com/AdaCore/gnatcoll-bindings/issues/19](https://github.com/AdaCore/gnatcoll-bindings/issues/19)

**Comment by Blady-Com** [comment]
Trouble with iconv again :-(
Concerning gnatcoll_iconv, last year, I was able to get out of it by building source codes with the appropriate options, thanks.

But this year, it depends on libstdc++.6.dylib that depends on libiconv from macOS:
```
% otool -L /Users/me/2025a/xnadalib-2025/lib/libstdc++.6.dylib
/Users/me/2025a/xnadalib-2025/lib/libstdc++.6.dylib:
	@rpath/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.34.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	@rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
% nm /Users/me/2025a/xnadalib-2025/lib/libstdc++.6.dylib | grep -i iconv
                 U _iconv
                 U _iconv_close
                 U _iconv_open
```
and not from GTK:
```
% nm /Users/me/2025a/xnadalib-2025/lib/libiconv.2.dylib | grep -i iconv 
0000000000124000 D __libiconv_version
0000000000006a00 T _iconv_canonicalize
00000000000054a0 T _libiconv
0000000000005530 T _libiconv_close
00000000000030f0 T _libiconv_open
```
In that case, how to hack without the source codes?

**Reply by jralls** [reply]
Why wouldn't you have the sources? This is a GNU project, the sources should all be on Savannah.

The first problem IMO is libstdc++. That's for GCC and on macOS we use LLVM whose C++ standard library is called libc++. Of course that also links the macOS libiconv, so it's orthogonal to your question.

The library paths confuse me. Does xnadalib-2025 provide both libstdc++ and libiconv? Maybe they screwed up their linkage?

Regardless of all of that if something you're building needs GNU libiconv then just make sure it appears in the linkage line *after* your object files and *before* libstdc++. The linker makes a table of unresolved symbols and then processes the libraries in the order presented on the command line, removing symbols as they're resolved and adding new ones from each library to be resolved as it goes down the list.

**Comment by Blady-Com** [comment]
> Why wouldn't you have the sources? This is a GNU project, the sources should all be on Savannah.

Well, I'm afraid of drowning too far.

> The first problem IMO is libstdc++. That's for GCC and on macOS we use LLVM whose C++ standard library is called libc++. Of course that also links the macOS libiconv, so it's orthogonal to your question.

It comes from GNAT Ada compiler based on GCC backend.

> The library paths confuse me. Does xnadalib-2025 provide both libstdc++ and libiconv? Maybe they screwed up their linkage?

In order it will be taken in account by GTK Mac Bundler, I copied it in my app root folder.
Thus when the app run I got:
```
dyld 24159: Symbol not found: _iconv
  Referenced from: <2DC5F646-9E49-3012-9933-92AF99EEF7DE> /Users/me/2025a/xnadalib-2025/lib/libstdc++.6.dylib
Expected in:     EBE92BA5-D441-309B-9022-FA9E905E7A3B /Users/me/2025a/xnadalib-2025/lib/libiconv.2.dylib
```

> Regardless of all of that if something you're building needs GNU libiconv then just make sure it appears in the linkage line _after_ your object files and _before_ libstdc++. The linker makes a table of unresolved symbols and then processes the libraries in the order presented on the command line, removing symbols as they're resolved and adding new ones from each library to be resolved as it goes down the list.

I'm not sure that I could change the order of linked libs as it comes from project files processed by a project builder.
However I can set [static linking mode](https://github.com/AdaCore/ada_language_server/blob/master/gnat/lsp_server.gpr#L69).
May this help ?

**Reply by jralls** [reply]
```
dyld 24159: Symbol not found: _iconv
  Referenced from: <2DC5F646-9E49-3012-9933-92AF99EEF7DE> /Users/me/2025a/xnadalib-2025/lib/libstdc++.6.dylib
Expected in:     EBE92BA5-D441-309B-9022-FA9E905E7A3B /Users/me/2025a/xnadalib-2025/lib/libiconv.2.dylib
```

Ah, I misunderstood the direction of the problem. I'm not sure that a static build of libstdc++ would do anything for you. Apple doesn't distribute a libiconv.a to statically link. In fact the SDK doesn't even have the dylib, it has libiconv.tbd that's Apple's equivalent to GNU libtool's libiconv.la: A text file that points to /usr/lib/libiconv.dylib that also doesn't actually exist. The machine code for most system libraries are in something called the dyld symbol cache that's buried in /System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/.

@techee submitted c338cc2 a couple of years ago because he had a problem with `iconv_ostream_create` in gettext on Sonoma. You might try a clean build (i.e. starting with bootstrap-gtk-osx) with after commenting out `<dep package="libiconv">` in the gettext module to see if it builds.

The other approach is to try to educate the packager of the GNAT Ada compiler that they need to use GNU libiconv to ensure compatibility with GNU gettext.

**Reply by Blady-Com** [reply]
> Ah, I misunderstood the direction of the problem. I'm not sure that a static build of libstdc++ would do anything for you. Apple doesn't distribute a libiconv.a to statically link. In fact the SDK doesn't even have the dylib, it has libiconv.tbd that's Apple's equivalent to GNU libtool's libiconv.la: A text file that points to /usr/lib/libiconv.dylib that also doesn't actually exist. The machine code for most system libraries are in something called the dyld symbol cache that's buried in /System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/.

That's give me an idea, I've prepended `/usr/lib` to `LD_LIBRARY_PATH`:

`export LD_LIBRARY_PATH=/usr/lib:"$bundle_lib:$bundle_lib/gdk-pixbuf-2.0/2.10.0/loaders"`

My app is happy with that :-)

> @techee submitted [c338cc2](https://github.com/jralls/gtk-osx-build/commit/c338cc23394571fda40e5c9037908e80e1961bf0) a couple of years ago because he had a problem with `iconv_ostream_create` in gettext on Sonoma. You might try a clean build (i.e. starting with bootstrap-gtk-osx) with after commenting out `<dep package="libiconv">` in the gettext module to see if it builds.

Thanks for the pointer. It's worth trying.
But not right away, as my app runs, I prefer to focus on next step that is bundling the app.

> The other approach is to try to educate the packager of the GNAT Ada compiler that they need to use GNU libiconv to ensure compatibility with GNU gettext.

Why not, but, frankly speaking, if the libiconv API have different names in order to coexist then the lib name should be different.
